lib/host/wasi/inode-win.cpp

```diff
@@ -687,18 +2146,102 @@ WasiExpect<void> INode::sockGetPeerAddr(uint8_t *, uint32_t *,
 }
 
 __wasi_filetype_t INode::unsafeFiletype() const noexcept {
+
+  // TODO: Find equivalents to the other file types
+  // To be completed along with other similar functions
+
+  if (unlikely(GetFileInformationByHandle(Handle, &(*FileInfo))) == FALSE) {
+    return __WASI_FILETYPE_UNKNOWN;
+  }
+
+  switch ((*FileInfo).dwFileAttributes) {
+  case FILE_ATTRIBUTE_DIRECTORY:
+    return __WASI_FILETYPE_DIRECTORY;
+  case FILE_ATTRIBUTE_NORMAL:
+    return __WASI_FILETYPE_REGULAR_FILE;
+  case FILE_ATTRIBUTE_REPARSE_POINT:
+    return __WASI_FILETYPE_SYMBOLIC_LINK;
+  }
+
+  switch (GetFileType(Handle)) {
+  case FILE_TYPE_CHAR:
+    return __WASI_FILETYPE_CHARACTER_DEVICE;
+  }
+  return __WASI_FILETYPE_UNKNOWN;
+
   return __WASI_FILETYPE_UNKNOWN;
 }
 
 WasiExpect<__wasi_filetype_t> INode::filetype() const noexcept {
-  return WasiUnexpect(__WASI_ERRNO_NOSYS);
+
+  // TODO: Find equivalents to the other file types
+  // To be completed along with other similar functions
+
+  if (unlikely(GetFileInformationByHandle(Handle, &(*FileInfo))) == FALSE) {
+    return WasiUnexpect(fromWinError(GetLastError()));
+  }
+
+  switch ((*FileInfo).dwFileAttributes) {
+  case FILE_ATTRIBUTE_DIRECTORY:
+    return __WASI_FILETYPE_DIRECTORY;
+  case FILE_ATTRIBUTE_NORMAL:
+    return __WASI_FILETYPE_REGULAR_FILE;
+  case FILE_ATTRIBUTE_REPARSE_POINT:
+    return __WASI_FILETYPE_SYMBOLIC_LINK;
+  }
+
+  switch (GetFileType(Handle)) {
+  case FILE_TYPE_CHAR:
+    return __WASI_FILETYPE_CHARACTER_DEVICE;
+  }
+  return __WASI_FILETYPE_UNKNOWN;
 }
 
-bool INode::isDirectory() const noexcept { return false; }
+bool INode::isDirectory() const noexcept {
+  if (FileInfo.has_value()) {
+    return (*FileInfo).dwFileAttributes == FILE_ATTRIBUTE_DIRECTORY;
+  }
 
-bool INode::isSymlink() const noexcept { return false; }
+  // Checking the if the Handle is valid or not because we cannot use
+  // WasiUnexpect() here
+  if (Handle != INVALID_HANDLE_VALUE &&
+      unlikely(GetFileInformationByHandle(Handle, &(*FileInfo))) == FALSE) {
+    return false;
+  }
 
-WasiExpect<__wasi_filesize_t> INode::filesize() const noexcept { return false; }
+  return (*FileInfo).dwFileAttributes == FILE_ATTRIBUTE_DIRECTORY;
+}
+
+bool INode::isSymlink() const noexcept {
+  if (FileInfo.has_value()) {
+    return (*FileInfo).dwFileAttributes == FILE_ATTRIBUTE_REPARSE_POINT;
+  }
+
+  // Checking the if the Handle is valid or not because we cannot use
+  // WasiUnexpect() here
+  if (Handle != INVALID_HANDLE_VALUE &&
+      unlikely(GetFileInformationByHandle(Handle, &(*FileInfo))) == FALSE) {
+    return false;
+  }
+
+  return (*FileInfo).dwFileAttributes == FILE_ATTRIBUTE_REPARSE_POINT;
+}
+
+WasiExpect<__wasi_filesize_t> INode::filesize() const noexcept {
+  if (FileInfo.has_value()) {
+    return (*FileInfo).dwFileAttributes == FILE_ATTRIBUTE_REPARSE_POINT;
```

Comment:
This function is `filesize`. Why do we return a condition check here?

Reply:
Must have copied the code from the function that checks for a symlink.